clis/teliod/src/nc.rs

```diff
@@ -0,0 +1,533 @@
+            app_user_uid: config.app_user_uid,
+            callbacks: callbacks.clone(),
+
+            http_certificate_file_path: config.http_certificate_file_path.clone(),
+            mqtt_certificate_file_path: config.mqtt_certificate_file_path.clone(),
+        };
+
+        start_mqtt(nc_config).await?;
+
+        Ok(Self { callbacks })
+    }
+
+    pub async fn add_callback(&self, callback: Callback) {
+        let mut callbacks = self.callbacks.lock().await;
+        callbacks.push(callback);
+    }
+}
+
+async fn start_mqtt(nc_config: NCConfig) -> Result<(), Error> {
+    let backoff_bounds = ExponentialBackoffBounds {
+        initial: Duration::from_secs(1),
+        maximal: Some(Duration::from_secs(300)),
+    };
+    let backoff = ExponentialBackoff::new(backoff_bounds)?;
+
+    tokio::spawn(async move {
+        loop {
+            let (client, mut eventloop, expires_at) =
+                Box::pin(connect_to_nc_with_backoff(&nc_config, backoff.clone())).await;
+
+            loop {
+                select! {
+                    _ = tokio::time::sleep_until(expires_at) => {
+                        info!("Notification Center credentials expired, mqtt will be restarted");
+                        break;
+                    },
+                    event = eventloop.poll() => {
+                        match event {
+                            Ok(Event::Incoming(Packet::Publish(p))) => {
+                                if let Err(e) = handle_incoming_publish(&client, p, &nc_config.callbacks).await {
+                                    error!("Failed to handle incoming publish: {e}");
+                                }
+                            }
+                            Ok(Event::Incoming(p)) => {
+                                debug!("mqtt incoming: {p:?}");
+                            }
+                            Ok(Event::Outgoing(p)) => debug!("mqtt outgoing event: {p:?}"),
+                            Err(e) => {
+                                error!("mqtt event loop error: {e}");
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    Ok(())
+}
+
+async fn connect_to_nc_with_backoff(
+    nc_config: &NCConfig,
+    mut backoff: ExponentialBackoff,
+) -> (AsyncClient, EventLoop, Instant) {
+    loop {
+        match connect_to_nc(nc_config).await {
+            Ok(mqtt) => return mqtt,
+            Err(e) => {
+                warn!(
+                    "Failed to connect to mqtt: {e}, will wait for {:?} and retry",
+                    backoff.get_backoff()
+                );
+                tokio::time::sleep(backoff.get_backoff()).await;
+                backoff.next_backoff();
+            }
+        }
+    }
+}
+
+async fn connect_to_nc(nc_config: &NCConfig) -> Result<(AsyncClient, EventLoop, Instant), Error> {
+    let credentials = request_nc_credentials(&nc_config).await?;
```

Comment:
Note that there is one failure path, which is a bit nasty...

If `nc` credentials are successfully received, but then MQTT connection fails - we should at least attempt to revoke them, as otherwise it is possible to pile them up on the server-side.

Reply:
I tried following scenario using `Postman` and `mosquitto_sub`:

1. Fetching NC credentials with POST `/v1/notifications/tokens`.
2. Subscribing to MQTT with fetched creds.
3. Sending PATCH request to `/v1/meshnet/machines`. Changes are seen in MQTT `/meshnet` topic.
4. Fetching NC credentials with POST `/v1/notifications/tokens` once again.
5. After that, I am _not_ disconnected from MQTT, but if I do disconnect manually, I am not able to use creds from step 2.

I am forcefully dropped from MQTT broker, if another connection with same identifier (and valid creds) joins in.

Reply:
So we're good there. If new creds are issued, the previous ones are revoked.

Reply:
@Jauler do you want me to remove the revocation?

Reply:
Yeah, I talked with the Notification Center devs, and they mentioned, that it is safe to issue multiple tokens with the same `app-user-id` meaning that my initial comment is not really relevant. 

I guess the original flow is much simpler, therefore maybe it makes sense to revert back?

Sorry for the confusion.

Reply:
Removed :)